### PR TITLE
Remove generated code from git

### PR DESCRIPTION
**Pull request by elliott-davis**
I am removing this after realizing that we are all checking in different version of code generated by different versions of protoc. After this change, the build will be the single source of truth for all protocol changes. Also, I have removed the need to run `cargo build --features protocols` as it now runs automatically.

Closes: https://github.com/habitat-sh/habitat/issues/4290

Signed-off-by: Elliott Davis <elliott@excellent.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by elliott-davis** [comment]
This is broken. Turns out we all have protofuf v3 installed and the hab package is still on v2. I'll take the time tomorrow to upgrade to 3.

**Comment by baumanj** [comment]
Does this resolve https://github.com/habitat-sh/habitat/issues/4290, @elliott-davis? If so, go ahead and self assign for your much deserved glory.

File: components/libbuild-protocols.rs [hunk 1]
```diff
@@ -48,7 +27,7 @@ mod protocols {
                 continue;
             }
             if file.metadata().unwrap().is_file() {
-                files.push(file.path());
+                files.push(file.path().to_string_lossy().into_owned());
```
Comment (baumanj) [inline components/libbuild-protocols.rs]: Seem like something has gone horribly wrong here if we somehow have invalid unicode in `file.path()`. I'd be more comfortable with
```rust
                  files.push(file.path().to_str().unwrap().into());
```

**Comment by elliott-davis** [comment]
@baumanj it does resolve that issue! I think it would be good to discuss the implications of this here or in that issue. Personally, I don't see a problem with the build server generating these protobuf files. It should uncover problems with dependencies changing (which is how I found this in the first place) since the build will break immediately.

File: components/builder-api/Cargo.toml [hunk 2]
```diff
@@ -12,11 +12,11 @@ path = "src/main.rs"
 doc = false
 
 [dependencies]
-clippy = {version = "*", optional = true}
-constant_time_eq = "*"
 base64 = "*"
 bodyparser = "*"
 builder-http-gateway = { path = "../builder-http-gateway" }
+clippy = {version = "*", optional = true}
+constant_time_eq = "*"
```
Comment (raskchanky) [inline components/builder-api/Cargo.toml]: Thank you.

File: components/libbuild-protocols.rs [hunk 3]
```diff
@@ -1,45 +1,24 @@
...
+        let protocols = protocol_files();
+        protoc_rust::run(protoc_rust::Args {
+            out_dir: "src/message",
+            input: protocols
+                .iter()
+                .map(AsRef::as_ref)
+                .collect::<Vec<&str>>()
+                .as_slice(),
+            includes: &["protocols"],
+        }).expect("protoc");
```
Comment (raskchanky) [inline components/libbuild-protocols.rs]: Perhaps the message that `expect` emits could be a little more descriptive?

**Comment by raskchanky** [comment]
So, just to make sure I understand:

* Currently (w/o this PR) a developer runs `cargo build --features protocols` and whatever version of `protoc` said developer has on their workstation reads in the `*.proto` files and spits out a bunch of Rust code in the `src/message` subdir, which then gets compiled into the rest of our code.  The files inside `src/message` are checked into git because w/o their presence, everything would break.
* With this PR, there's no longer a need to run `cargo build --features protocols` because the protocols get re-compiled automatically every time `cargo build` is run.  `*.proto` files are still read in and Rust code is still emitted to the `src/message` subdir, just like before, with the main difference being that code is now git ignored and we use a library to generate it instead of a `protoc` binary.

Do I have that correct?

If I do, then this seems like a great change.  One downside I see off the cuff is an increased compile time for the `builder-protocols` crate, since it has to re-compile all the proto definitions every time, instead of only when they change.  Honestly, though, this crate doesn't change frequently and the increased convenience (not to mention correctness from not having to deal with different versions of `protoc`) feels worth it.  And, if someone really felt like the speed bump was onerous, you could probably implement some crude make-like logic in `build.rs` to check file timestamps and conditionally compile based on that.

**Comment by elliott-davis** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/builder/builds/357423159).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/builder/builds/357423159).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Comment by chefsalim** [comment]
One thing that we're missing (that may be orthogonal), is how to enable a simple "cargo check" outside of the hab build?  I think we should continue to support that..

**Comment by raskchanky** [comment]
@chefsalim I don't think I understand.  What's preventing `cargo check` from working with this as is?

**Comment by raskchanky** [comment]
Kinda weird, but it looks like thesentinels didn't merge this.  I'll merge manually.

**Comment by chefsalim** [comment]
@raskchanky there are no instructions or automation on how to enable a protocols build outside of the studio build - it's not this PR that's the cause, but in general we've lost that capability...

**Comment by elliott-davis** [comment]
@chefsalim `cargo check` now automatically generates the proto messages :tada: 

As far as running cargo inside the dev studio, there is still some work to be done.

**Comment by raskchanky** [comment]
image

@chefsalim Maybe I'm missing something here?  Seems fine to me.

**Comment by raskchanky** [comment]
image

Full `cargo build` seems to work too.

**Comment by raskchanky** [comment]
Note that both screenshots above were done outside the studio @chefsalim

**Comment by chefsalim** [comment]
Right, but if you spin up a new VM, it's not going to work - you're not going to have any protobuf libraries installed..

**Comment by raskchanky** [comment]
I don't think this PR removed any protobuf libraries, though, did it?  I don't see that in the diff.  So unless I'm misunderstanding, that would've also been a problem before this PR got merged?

**Comment by chefsalim** [comment]
Yes, that's correct - it's an orthogonal issue, this PR is not the cause

**Comment by chefsalim** [comment]
Opened new issue to track the above discussion https://github.com/habitat-sh/builder/issues/344

**Pull request merged**